File: sdk/swift/Sources/PrivacyRPCSDK/PhishingDetector.swift
import Foundation

/// Phishing Detector for PrivacyRPC
///
/// Local-only phishing detection with homograph attack detection,
/// typosquatting analysis, and known domain verification.
/// Matches the Chrome extension and TypeScript SDK phishing module.
public class PhishingDetector {

    // MARK: - Models

    public struct PhishingResult {
        public let domain: String
        public let isPhishing: Bool
        public let confidence: Confidence
        public let reason: String?
        public let legitimateDomain: String?
        public let alerts: [String]

        public enum Confidence: String {
            case confirmed = "CONFIRMED"
            case high = "HIGH"
            case medium = "MEDIUM"
            case low = "LOW"
            case unknown = "UNKNOWN"
        }
    }

    // MARK: - Known Legitimate Domains

    private static let legitimateDomains: Set<String> = [
        // Wallets
        "phantom.app", "solflare.com", "backpack.app", "glow.app",
        // DEXs
        "jup.ag", "jupiter.ag", "raydium.io", "orca.so", "lifinity.io",
        "meteora.ag", "phoenix.trade", "drift.trade", "zeta.markets", "pump.fun",
        // DeFi
        "marinade.finance", "solend.fi", "mango.markets", "kamino.finance",
        "marginfi.com", "solblaze.org", "jito.network",
        // NFT
        "magiceden.io", "tensor.trade", "hyperspace.xyz", "exchange.art",
        "formfunction.xyz", "solanart.io", "opensea.io",
        // Explorers
        "solana.com", "solscan.io", "solanabeach.io", "solana.fm",
        // Infrastructure
        "helius.dev", "shyft.to", "quicknode.com", "alchemy.com",
        // Other
        "squads.so", "realms.today", "dialect.to"
    ]

    /// Known phishing domains
    private static let knownPhishing: Set<String> = [
        "phantom-wallet.app", "phantomm.app", "solanaa.com",
        "jupiterr.ag", "magiceden.xyz", "pump.fun-watch.dev"
    ]

    /// Homograph character mappings (Unicode lookalikes)
    private static let homoglyphs: [Character: [Character]] = [
        "a": ["а", "ɑ", "α"],  // Cyrillic a, Latin alpha, Greek alpha
        "e": ["е", "ε", "ё"],  // Cyrillic ie, Greek epsilon
        "o": ["о", "ο", "0"],  // Cyrillic o, Greek omicron, zero
        "p": ["р", "ρ"],       // Cyrillic er, Greek rho
        "c": ["с", "ϲ"],       // Cyrillic es, Greek lunate sigma
        "x": ["х", "χ"],       // Cyrillic ha, Greek chi
        "s": ["ѕ"],            // Cyrillic dze
        "i": ["і", "ι", "1"],  // Cyrillic i, Greek iota, one
        "n": ["п"],            // Cyrillic en
        "t": ["т"],            // Cyrillic te
        "y": ["у"],            // Cyrillic u
    ]

    public init() {}

    // MARK: - Public API

    /// Check a domain for phishing indicators
    public func check(_ domain: String) -> PhishingResult {
        let normalized = domain.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        var alerts: [String] = []

        // 1. Known phishing domain
        if Self.knownPhishing.contains(normalized) {
            return PhishingResult(
                domain: normalized,
                isPhishing: true,
                confidence: .confirmed,
                reason: "Known phishing domain",
                legitimateDomain: nil,
                alerts: ["This domain is on the known phishing list"]
            )
        }

        // 2. Known legitimate domain
        if Self.legitimateDomains.contains(normalized) {
            return PhishingResult(
                domain: normalized,
                isPhishing: false,
                confidence: .confirmed,
                reason: nil,
                legitimateDomain: normalized,
                alerts: []
            )
        }

        // 3. Homograph detection
        if let homographResult = detectHomograph(normalized) {
            return homographResult
        }

        // 4. Typosquatting detection
        if let typoResult = detectTyposquatting(normalized) {
            return typoResult
        }

        // 5. Suspicious pattern detection
        let patternAlerts = detectSuspiciousPatterns(normalized)
        alerts.append(contentsOf: patternAlerts)

        if !alerts.isEmpty {
            return PhishingResult(
                domain: normalized,
                isPhishing: false,
                confidence: .low,
                reason: "Some suspicious patterns detected",
                legitimateDomain: nil,
                alerts: alerts
            )
        }

        // Unknown domain
        return PhishingResult(
            domain: normalized,
            isPhishing: false,
            confidence: .unknown,
            reason: nil,
            legitimateDomain: nil,
            alerts: []
        )
    }

    // MARK: - Detection Methods

    private func detectHomograph(_ domain: String) -> PhishingResult? {
        // Check if domain contains non-ASCII characters
        let hasNonAscii = domain.unicodeScalars.contains { !$0.isASCII }
        guard hasNonAscii else { return nil }

        // Try to normalize to ASCII and check against known domains
        let asciiVersion = normalizeToAscii(domain)

        for legit in Self.legitimateDomains {
            if asciiVersion == legit || levenshteinDistance(asciiVersion, legit) <= 1 {
                return PhishingResult(
                    domain: domain,
                    isPhishing: true,
                    confidence: .high,
                    reason: "Homograph attack detected - uses lookalike characters to impersonate \(legit)",
                    legitimateDomain: legit,
                    alerts: ["Domain contains non-ASCII lookalike characters"]
                )
            }
        }

        return PhishingResult(
            domain: domain,
            isPhishing: false,
            confidence: .medium,
            reason: "Contains non-ASCII characters",
            legitimateDomain: nil,
            alerts: ["Domain contains unusual Unicode characters"]
        )
    }

    private func detectTyposquatting(_ domain: String) -> PhishingResult? {
        let domainBase = domain.components(separatedBy: ".").first ?? domain

        for legit in Self.legitimateDomains {
            let legitBase = legit.components(separatedBy: ".").first ?? legit

            let distance = levenshteinDistance(domainBase, legitBase)

            // Very similar to a legitimate domain (1-2 character difference)
            if distance == 1 {
                return PhishingResult(
                    domain: domain,
                    isPhishing: true,
                    confidence: .high,
                    reason: "Possible typosquatting of \(legit) (edit distance: \(distance))",
                    legitimateDomain: legit,
                    alerts: ["Very similar to known legitimate domain: \(legit)"]
                )
            } else if distance == 2 && domainBase.count > 4 {
                return PhishingResult(
                    domain: domain,
                    isPhishing: false,
                    confidence: .medium,
                    reason: "Somewhat similar to \(legit)",
                    legitimateDomain: legit,
                    alerts: ["Resembles legitimate domain: \(legit)"]
                )
            }
        }

        return nil
    }

    private func detectSuspiciousPatterns(_ domain: String) -> [String] {
        var alerts: [String] = []

        // Check for wallet/crypto keywords combined with suspicious TLDs
        let suspiciousKeywords = ["wallet", "phantom", "solana", "crypto", "airdrop", "claim", "free"]
        let suspiciousTLDs = [".xyz", ".site", ".online", ".top", ".buzz", ".click"]

        let hasKeyword = suspiciousKeywords.contains { domain.contains($0) }
        let hasSuspiciousTLD = suspiciousTLDs.contains { domain.hasSuffix($0) }

        if hasKeyword && hasSuspiciousTLD {
            alerts.append("Crypto keyword with suspicious TLD")
        }

        // Check for hyphenated versions of known domains
        for legit in Self.legitimateDomains {
            let legitBase = legit.components(separatedBy: ".").first ?? legit
            if domain.contains("\(legitBase)-") || domain.contains("-\(legitBase)") {
                alerts.append("Contains hyphenated version of known domain: \(legit)")
            }
        }

        return alerts
    }

    // MARK: - Helpers

    private func normalizeToAscii(_ text: String) -> String {
        var result = ""
        for char in text {
            var replaced = false
            for (ascii, lookalikes) in Self.homoglyphs {
                if lookalikes.contains(char) {
                    result.append(ascii)
                    replaced = true
                    break
                }
            }
            if !replaced {
                result.append(char)
            }
        }
        return result
    }

    private func levenshteinDistance(_ s1: String, _ s2: String) -> Int {
        let a = Array(s1)
        let b = Array(s2)
        let m = a.count
        let n = b.count

        if m == 0 { return n }
        if n == 0 { return m }

        var matrix = Array(repeating: Array(repeating: 0, count: n + 1), count: m + 1)

        for i in 0...m { matrix[i][0] = i }
        for j in 0...n { matrix[0][j] = j }

        for i in 1...m {
            for j in 1...n {
                let cost = a[i - 1] == b[j - 1] ? 0 : 1
                matrix[i][j] = min(
                    matrix[i - 1][j] + 1,
                    matrix[i][j - 1] + 1,
                    matrix[i - 1][j - 1] + cost
                )
            }
        }

        return matrix[m][n]
    }
}

File: sdk/kotlin/src/main/kotlin/com/privacyrpc/sdk/PhishingDetector.kt
package com.privacyrpc.sdk

/**
 * Phishing/Spoofing Website Detector
 *
 * Detects fake crypto websites using:
 * 1. Homograph attacks (phantóm.app vs phantom.app)
 * 2. Typosquatting (phantm.app, phanton.app)
 * 3. Domain lookalikes (phantom-app.com, phantom.io)
 * 4. Known phishing domains blacklist
 * 5. Certificate verification for legitimate sites
 *
 * This is where most wallet drains actually happen - users
 * visit fake sites and enter their seed phrase or sign malicious transactions.
 */
class PhishingDetector(
    private val alertHandler: AlertHandler? = null
) {
    companion object {
        // Homograph characters that look like ASCII
        private val HOMOGRAPHS = mapOf(
            'а' to 'a', // Cyrillic
            'е' to 'e',
            'о' to 'o',
            'р' to 'p',
            'с' to 'c',
            'у' to 'y',
            'х' to 'x',
            'ѕ' to 's',
            'і' to 'i',
            'ј' to 'j',
            'һ' to 'h',
            'ԁ' to 'd',
            'ԝ' to 'w',
            'ɑ' to 'a', // Latin Extended
            'ɡ' to 'g',
            'ɩ' to 'i',
            'ɪ' to 'i',
            'ι' to 'i', // Greek
            'ο' to 'o',
            'α' to 'a',
            'ν' to 'v',
            'τ' to 't',
            '0' to 'o', // Numbers
            '1' to 'l',
            '5' to 's',
            '3' to 'e',
        )
    }

    // Legitimate crypto domains and their info
    private val legitimateDomains = mapOf(
        // Wallets
        "phantom.app" to DomainInfo("Phantom Wallet", DomainCategory.WALLET),
        "backpack.app" to DomainInfo("Backpack Wallet", DomainCategory.WALLET),
        "solflare.com" to DomainInfo("Solflare Wallet", DomainCategory.WALLET),
        "metamask.io" to DomainInfo("MetaMask Wallet", DomainCategory.WALLET),
        "trustwallet.com" to DomainInfo("Trust Wallet", DomainCategory.WALLET),

        // DEXes
        "jup.ag" to DomainInfo("Jupiter Exchange", DomainCategory.DEX),
        "jupiter.ag" to DomainInfo("Jupiter Exchange", DomainCategory.DEX),
        "raydium.io" to DomainInfo("Raydium", DomainCategory.DEX),
        "orca.so" to DomainInfo("Orca", DomainCategory.DEX),
        "uniswap.org" to DomainInfo("Uniswap", DomainCategory.DEX),
        "pump.fun" to DomainInfo("Pump.fun", DomainCategory.DEX),

        // Solana ecosystem
        "solana.com" to DomainInfo("Solana", DomainCategory.BLOCKCHAIN),
        "solscan.io" to DomainInfo("Solscan Explorer", DomainCategory.EXPLORER),
        "solana.fm" to DomainInfo("Solana FM Explorer", DomainCategory.EXPLORER),
        "helius.dev" to DomainInfo("Helius", DomainCategory.RPC),
        "helius.xyz" to DomainInfo("Helius", DomainCategory.RPC),
        "quicknode.com" to DomainInfo("QuickNode", DomainCategory.RPC),

        // NFT marketplaces
        "magiceden.io" to DomainInfo("Magic Eden", DomainCategory.NFT),
        "tensor.trade" to DomainInfo("Tensor", DomainCategory.NFT),
        "opensea.io" to DomainInfo("OpenSea", DomainCategory.NFT),

        // Other
        "coingecko.com" to DomainInfo("CoinGecko", DomainCategory.DATA),
        "coinmarketcap.com" to DomainInfo("CoinMarketCap", DomainCategory.DATA),
        "dexscreener.com" to DomainInfo("DEX Screener", DomainCategory.DATA),
        "birdeye.so" to DomainInfo("Birdeye", DomainCategory.DATA),
    )

    // Known phishing domains (would be updated regularly in production)
    private val knownPhishingDomains = mutableSetOf(
        "phantom-wallet.app",
        "phantomwallet.io",
        "phantom-app.com",
        "solana-airdrop.com",
        "jupiter-airdrop.com",
        "free-solana.com",
        "claim-jupiter.com",
        // pump.fun phishing
        "pump.fun-watch.dev",
    )

    /**
     * Check if a domain is potentially a phishing site
     */
    fun checkDomain(domain: String): PhishingResult {
        val normalizedDomain = normalizeDomain(domain)

        // Check known phishing list
        if (normalizedDomain in knownPhishingDomains) {
            return PhishingResult(
                isPhishing = true,
                confidence = Confidence.CONFIRMED,
                reason = "Known phishing domain",
                legitimateDomain = findSimilarLegitimate(normalizedDomain),
                alerts = listOf("This is a confirmed phishing site. Do NOT enter any information.")
            ).also { notifyAlert(it, domain) }
        }

        // Check if it's a legitimate domain
        if (normalizedDomain in legitimateDomains) {
            return PhishingResult(
                isPhishing = false,
                confidence = Confidence.CONFIRMED,
                reason = "Verified legitimate domain",
                legitimateDomain = normalizedDomain
            )
        }

        // Check for homograph attacks
        val homographResult = checkHomograph(normalizedDomain)
        if (homographResult != null) {
            return homographResult.also { notifyAlert(it, domain) }
        }

        // Check for typosquatting
        val typoResult = checkTyposquatting(normalizedDomain)
        if (typoResult != null) {
            return typoResult.also { notifyAlert(it, domain) }
        }

        // Check for suspicious patterns
        val patternResult = checkSuspiciousPatterns(normalizedDomain)
        if (patternResult != null) {
            return patternResult.also { notifyAlert(it, domain) }
        }

        // Unknown domain - could be legitimate, could be phishing
        return PhishingResult(
            isPhishing = false,
            confidence = Confidence.UNKNOWN,
            reason = "Unknown domain - exercise caution",
            legitimateDomain = null
        )
    }

    /**
     * Check for homograph (lookalike character) attacks
     */
    private fun checkHomograph(domain: String): PhishingResult? {
        // Convert homographs to ASCII
        val asciiDomain = domain.map { HOMOGRAPHS[it] ?: it }.joinToString("")

        // If conversion changed the domain, it contains homographs
        if (asciiDomain != domain) {
            val matchedLegit = legitimateDomains.keys.find { it == asciiDomain }
            if (matchedLegit != null) {
                return PhishingResult(
                    isPhishing = true,
                    confidence = Confidence.HIGH,
                    reason = "Homograph attack detected - uses lookalike characters",
                    legitimateDomain = matchedLegit,
                    alerts = listOf(
                        "This domain uses fake characters to look like '$matchedLegit'",
                        "Example: 'а' (Cyrillic) looks like 'a' (Latin) but is different",
                        "The real site is: $matchedLegit"
                    )
                )
            }
        }

        return null
    }

    /**
     * Check for typosquatting (common typos)
     */
    private fun checkTyposquatting(domain: String): PhishingResult? {
        for (legitDomain in legitimateDomains.keys) {
            val distance = levenshteinDistance(domain, legitDomain)

            // Very close match (1-2 character difference)
            if (distance in 1..2) {
                return PhishingResult(
                    isPhishing = true,
                    confidence = Confidence.HIGH,
                    reason = "Typosquatting detected - similar to legitimate domain",
                    legitimateDomain = legitDomain,
                    alerts = listOf(
                        "This looks like a typo of '$legitDomain'",
                        "Did you mean to visit: $legitDomain?"
                    )
                )
            }

            // Somewhat close (3 character difference) - warn but lower confidence
            if (distance == 3 && domain.length > 6) {
                return PhishingResult(
                    isPhishing = true,
                    confidence = Confidence.MEDIUM,
                    reason = "Possible typosquatting - similar to legitimate domain",
                    legitimateDomain = legitDomain,
                    alerts = listOf(
                        "This domain is similar to '$legitDomain'",
                        "Please verify you're on the correct site"
                    )
                )
            }
        }

        return null
    }

    /**
     * Check for suspicious URL patterns
     */
    private fun checkSuspiciousPatterns(domain: String): PhishingResult? {
        val suspiciousPatterns = listOf(
            // Fake wallet patterns
            Regex("phantom[^a-z]") to "phantom.app",
            Regex("metamask[^a-z]") to "metamask.io",
            Regex("solflare[^a-z]") to "solflare.com",
            Regex("backpack[^a-z]") to "backpack.app",

            // Fake DEX patterns
            Regex("jupiter[^a-z]") to "jup.ag",
            Regex("raydium[^a-z]") to "raydium.io",
            Regex("uniswap[^a-z]") to "uniswap.org",

            // Airdrop scams
            Regex("(airdrop|claim|free).*(solana|sol|jupiter|jup)") to null,
            Regex("(solana|sol|jupiter|jup).*(airdrop|claim|free)") to null,

            // Wallet connect scams
            Regex("wallet.?connect.?(verify|validate|sync)") to null,
            Regex("(verify|validate|sync).?wallet") to null,

            // Recovery scams
            Regex("(recover|restore).*(wallet|phrase|seed)") to null,
        )

        for ((pattern, legitDomain) in suspiciousPatterns) {
            if (pattern.containsMatchIn(domain)) {
                return PhishingResult(
                    isPhishing = true,
                    confidence = Confidence.HIGH,
                    reason = "Suspicious domain pattern detected",
                    legitimateDomain = legitDomain,
                    alerts = listOf(
                        "This domain matches known phishing patterns",
                        if (legitDomain != null) "The real site is: $legitDomain" else "This is likely a scam site",
                        "NEVER enter your seed phrase on any website"
                    )
                )
            }
        }

        return null
    }

    /**
     * Find the most similar legitimate domain
     */
    private fun findSimilarLegitimate(domain: String): String? {
        return legitimateDomains.keys
            .map { it to levenshteinDistance(domain, it) }
            .filter { it.second <= 5 }
            .minByOrNull { it.second }
            ?.first
    }

    /**
     * Calculate Levenshtein distance between two strings
     */
    private fun levenshteinDistance(s1: String, s2: String): Int {
        val dp = Array(s1.length + 1) { IntArray(s2.length + 1) }

        for (i in 0..s1.length) dp[i][0] = i
        for (j in 0..s2.length) dp[0][j] = j

        for (i in 1..s1.length) {
            for (j in 1..s2.length) {
                dp[i][j] = if (s1[i - 1] == s2[j - 1]) {
                    dp[i - 1][j - 1]
                } else {
                    minOf(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
                }
            }
        }

        return dp[s1.length][s2.length]
    }

    /**
     * Normalize domain for comparison
     */
    private fun normalizeDomain(domain: String): String {
        return domain
            .lowercase()
            .removePrefix("www.")
            .removePrefix("https://")
            .removePrefix("http://")
            .split("/")[0]  // Remove path
    }

    /**
     * Add a known phishing domain
     */
    fun reportPhishing(domain: String) {
        knownPhishingDomains.add(normalizeDomain(domain))
    }

    /**
     * Add a legitimate domain
     */
    fun addLegitimate(domain: String, name: String, category: DomainCategory) {
        // Note: In production, this would be admin-only or verified
    }

    /**
     * Notify alert handler
     */
    private fun notifyAlert(result: PhishingResult, domain: String) {
        if (result.isPhishing && alertHandler != null) {
            alertHandler.onAlert(PrivacyRPCAlert(
                type = AlertType.PHISHING_DETECTED,
                severity = when (result.confidence) {
                    Confidence.CONFIRMED -> Severity.CRITICAL
                    Confidence.HIGH -> Severity.CRITICAL
                    Confidence.MEDIUM -> Severity.HIGH
                    Confidence.LOW -> Severity.MEDIUM
                    Confidence.UNKNOWN -> Severity.LOW
                },
                message = result.reason,
                hostname = domain,
                details = mapOf(
                    "phishing_domain" to domain,
                    "legitimate_domain" to (result.legitimateDomain ?: "unknown"),
                    "confidence" to result.confidence.name,
                    "alerts" to result.alerts
                )
            ))
        }
    }
}

/**
 * Result of phishing check
 */
data class PhishingResult(
    val isPhishing: Boolean,
    val confidence: Confidence,
    val reason: String,
    val legitimateDomain: String?,
    val alerts: List<String> = emptyList()
)

/**
 * Confidence level of detection
 */
enum class Confidence {
    CONFIRMED,  // Known phishing domain
    HIGH,       // Very likely phishing
    MEDIUM,     // Probably phishing
    LOW,        // Possibly phishing
    UNKNOWN     // Can't determine
}

/**
 * Category of legitimate domain
 */
enum class DomainCategory {
    WALLET,
    DEX,
    NFT,
    BLOCKCHAIN,
    EXPLORER,
    RPC,
    DATA,
    OTHER
}

/**
 * Info about a legitimate domain
 */
data class DomainInfo(
    val name: String,
    val category: DomainCategory
)
